main.go

```diff
@@ -579,3 +582,29 @@ func setLoggerForProduction(encoder zapcore.LevelEncoder, dest io.Writer) {
 	ctrl.SetLogger(newlogger)
 	klog.SetLogger(newlogger)
 }
+
+// setupSignalHandler registers a signal handler for SIGTERM and SIGINT and
+// returns a context which is canceled when one of these signals is caught after
+// waiting for the specified period. If a second signal is caught then we
+// terminate immediately with exit code 1. The implementation has been stolen
+// from controller-runtime and then extended with wait period support:
+// https://github.com/kubernetes-sigs/controller-runtime/blob/2154ffbc22e26ffd8c3b713927f0df2fa40841f2/pkg/manager/signals/signal.go#L27-L45
+func setupSignalHandler(waitPeriod time.Duration) context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		// Cancel the context once the wait period expires but avoid blocking if
+		// a second signal is received
+		go func() {
+			<-time.After(waitPeriod)
```

Comment:
@dippynark, I'm not quite understanding the intent of this change because cancelling the context is the only way to signal to controller-runtime to shutdown.

Are you saying that there are other components within Gatekeeper that intercept the OS signal directly as well instead of relying on that parent context associated with the OS signal and it's child contexts? If not, then I read this code as sleeping for 10 seconds after the exit signal is received and then it instructs controller-runtime to shutdown, which just delays the existing behavior by 10 seconds.

Reply:
@mprahl The intent is to wait for some time after the TERM signal is received to get the following behaviour:

1. TERM signal is sent to the Gatekeeper process by the kubelet
2. API server stops sending new connections to Gatekeeper
3. Gatekeeper shuts down webhook server listeners

Currently step 2 can come after step 3 if we are unlucky (i.e. there is a race condition) so this change is to try to ensure that step 3 always comes after step 2 by delaying shutdown for a short amount of time. The OPA server PR linked from this PR description was for the same reason

Reply:
Got it. That makes sense to me now.